### ISSUE TITLE
All exported service functions are rebound

### ISSUE THREAD

**Issue by mdarveau** [balderdashy/sails#3005]
Since https://github.com/balderdashy/sails/commit/97861aa30da56eef2907516010ac31157e8c3abe, all exported service functions are bound to the service. It causes an unexpected side effect when upgrading from sails 1.10.x to 1.11.x.

I am using `CoffeeScript` and `Bluebird`. I declare an exception class in `UserService`:
```
exports.AuthenticationRequired = class AuthenticationRequired extends Error
  constructor: () ->
    @message = "Authentication required"
    @name = "AuthenticationRequired"
    Error.captureStackTrace(this, AuthenticationRequired)
```

And then throw it, also in `UserService`:
```
# Return the authenticated user or throw AuthenticationRequired
exports.doSomething = ( req ) ->
   ...
   throw new AuthenticationRequired() unless authenticated
   ...
```

And the controller does something like:
```
Bluebird.try ()->
   UserService.doSomething(req)
 .then ( user ) ->
    return res.json( user )
.catch UserService.AuthenticationRequired, ()->
   res.status(401).json( { error: 'Authentication required'} )
```

The problem is that `Bluebird` expects `AuthenticationRequired.prototype instanceof Error`, which it is when checked in the `UserService`. The problem is that outside of the service (in the controller for example), `AuthenticationRequired.prototype` is not an instance of `Error` since the exposed `AuthenticationRequired` was bound to `UserService` and is now a bind function wrapper around the original `AuthenticationRequired`.

Am I doing something wrong here? Am I supposed to expose the `AuthenticationRequired` constructor another way?

**Comment by tjwebb** [comment]
Your services should be plain objects, not classes. I have no idea what CoffeeScript is doing under the hood, since I've never used CoffeeScript, so I can't really diagnose what exactly the issue is.

**Comment by mdarveau** [comment]
The Coffeescript `class` is simply a utility to create a constructor and assign some methods to it's prototype.

This Coffeescript:
```
class Animal
  constructor: (@name) ->

  move: (meters) ->
    alert @name + " moved #{meters}m."
```

Will translate to:
```
Animal = (function() {
  function Animal(name) {
    this.name = name;
  }

  Animal.prototype.move = function(meters) {
    return alert(this.name + (" moved " + meters + "m."));
  };

  return Animal;

})();
```

In my case, it is that constructor function that I want to export as one attribute of my service. The problem is that the function is modified by the sails framework via `bind` and it changes the `.prototype` attribute on which `Bluebird` rely.

Any chance you rollback https://github.com/balderdashy/sails/commit/97861aa30da56eef2907516010ac31157e8c3abe?

**Comment by tjwebb** [comment]
> The Coffeescript class is simply a utility to create a constructor and assign some methods to it's prototype.
>> function Animal(name) {
    this.name = name;
  }

Right, the result will be a `Function` not an `Object`. Sails Services (that is, things in api/services) should be objects. There are some requests to support prototypal inheritance in Services, but that feature has not yet been implemented.

**Comment by mdarveau** [comment]
There must be a subtility I don't get. Event if I export an `Object`, all the functions of that object are still rebounded. Let say my service is:
```
module.exports =

  AuthenticationRequired: class AuthenticationRequired extends Error
    constructor: () ->
      @message = "Authentication required"
      @name = "AuthenticationRequired"
      Error.captureStackTrace(this, AuthenticationRequired)

  hashPassword: ( user ) ->
    ...
```

The `AuthenticationRequired` and the `hashPassword` function are rebound by the `_.bindAll(module)` that was introduced in https://github.com/balderdashy/sails/commit/97861aa30da56eef2907516010ac31157e8c3abe.

**Comment by tjwebb** [comment]
>   AuthenticationRequired: class AuthenticationRequired extends Error

`class` will export a function, not an object.

**Comment by sailsbot** [comment]
Thanks for posting, @mdarveau.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
